kaldb/src/main/java/com/slack/kaldb/metadata/core/KaldbMetadataStore.java

```diff
@@ -1,176 +1,200 @@
-import static com.google.common.base.Preconditions.checkState;
 import static com.slack.kaldb.server.KaldbConfig.DEFAULT_ZK_TIMEOUT_SECS;
 
-import com.google.common.base.Function;
-import com.google.common.util.concurrent.Futures;
-import com.google.common.util.concurrent.ListenableFuture;
-import com.google.common.util.concurrent.MoreExecutors;
-import com.google.protobuf.InvalidProtocolBufferException;
-import com.slack.kaldb.metadata.zookeeper.InternalMetadataStoreException;
-import com.slack.kaldb.metadata.zookeeper.MetadataStore;
-import com.slack.kaldb.metadata.zookeeper.NodeExistsException;
-import java.util.ArrayList;
-import java.util.Collections;
+import java.io.Closeable;
 import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
-import javax.annotation.Nullable;
-import org.apache.curator.utils.ZKPaths;
-import org.slf4j.Logger;
+import java.util.stream.Collectors;
+import org.apache.curator.x.async.AsyncCuratorFramework;
+import org.apache.curator.x.async.AsyncStage;
+import org.apache.curator.x.async.api.CreateOption;
+import org.apache.curator.x.async.modeled.ModelSerializer;
+import org.apache.curator.x.async.modeled.ModelSpec;
+import org.apache.curator.x.async.modeled.ModeledFramework;
+import org.apache.curator.x.async.modeled.ZPath;
+import org.apache.curator.x.async.modeled.cached.CachedModeledFramework;
+import org.apache.curator.x.async.modeled.cached.ModeledCacheListener;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.data.Stat;
 
 /**
- * KaldbMetadataStore is an abstract class on top of which all the metadata stores will be built.
+ * KaldbMetadataStore is a class which provides consistent ZK apis for all the metadata store class.
  *
- * <p>This abstraction is needed so that we can limit the ZK access to the application to a specific
- * set of paths and also to limit the operations that can be performed on those nodes. For example,
- * we only want the SnapshotMetadata to ever be created or deleted but never updated.
- *
- * <p>Every method provides an async(returns a future) and a sync API. In general, use the async API
- * you are performing batch operations and a sync if you are performing a synchronous operation on a
- * node.
+ * <p>Every method provides an async and a sync API. In general, use the async API you are
+ * performing batch operations and a sync if you are performing a synchronous operation on a node.
  */
-abstract class KaldbMetadataStore<T extends KaldbMetadata> {
-  protected final MetadataStore metadataStore;
-  protected final String storeFolder;
-  protected final MetadataSerializer<T> metadataSerializer;
-  protected final Logger logger;
+public class KaldbMetadataStore<T extends KaldbMetadata> implements Closeable {
```

Comment:
I think my biggest feedback from this class is, can this class return CompletableFuture instead of AsyncStage for the async operations? that way we still keep the ZK exposure minimum outside of this class?

Reply:
I think that we could return a `CompletionStage` here instead of an `AsyncStage` - that way it's part of the java concurrent package, isn't tied to Curator (like AsyncStage), and doesn't requiring converting these all to `CompleteableFutures`, since `AsyncStage` just extends `CompletionStage`.

Reply:
Made these all return `CompletionStage.`